tests/suites/test_suite_x509write.function

```diff
@@ -95,6 +167,48 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE depends_on:MBEDTLS_PEM_WRITE_C:MBEDTLS_X509_CSR_WRITE_C:MBEDTLS_USE_PSA_CRYPTO */
+void x509_csr_check_opaque( char * key_file, int md_type, int key_usage,
+                                 int cert_type )
+{
+    mbedtls_pk_context key;
+    mbedtls_x509write_csr req;
+    unsigned char buf[4096];
+    int ret;
+    size_t pem_len = 0;
+    const char *subject_name = "C=NL,O=PolarSSL,CN=PolarSSL Server 1";
+    rnd_pseudo_info rnd_info;
+
+    psa_crypto_init();
+    memset( &rnd_info, 0x2a, sizeof( rnd_pseudo_info ) );
+
+    mbedtls_pk_init( &key );
+    TEST_ASSERT( mbedtls_pk_parse_keyfile( &key, key_file, NULL ) == 0 );
+    TEST_ASSERT( wrap_key_as_opaque( &key, PSA_ALG_SHA_1 ) == 0 );
+
+    mbedtls_x509write_csr_init( &req );
+    mbedtls_x509write_csr_set_md_alg( &req, md_type );
+    mbedtls_x509write_csr_set_key( &req, &key );
+    TEST_ASSERT( mbedtls_x509write_csr_set_subject_name( &req, subject_name ) == 0 );
+    if( key_usage != 0 )
+        TEST_ASSERT( mbedtls_x509write_csr_set_key_usage( &req, key_usage ) == 0 );
+    if( cert_type != 0 )
+        TEST_ASSERT( mbedtls_x509write_csr_set_ns_cert_type( &req, cert_type ) == 0 );
+
+    ret = mbedtls_x509write_csr_pem( &req, buf, sizeof( buf ),
+                             rnd_pseudo_rand, &rnd_info );
+    TEST_ASSERT( ret == 0 );
+
+    pem_len = strlen( (char *) buf );
+    buf[pem_len] = '\0';
+    TEST_ASSERT( x509_crt_verifycsr(buf, pem_len+1, &key) == 0 );
```

Comment:
Major: Why does this test deviate from the existing CSR writing tests in that it attempts a verification instead of a binary `memcmp()`? This (a) introduces the need to support opaque verification here which could otherwise be omitted for the benefit of simplicity of this PR, and (b) Introduces a dependency on X.509 parsing which doesn't exist anywhere else in the test suite.

@mpg @AndrzejKurek I'd suggest to adapt the code from https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_x509write.function#L39 to allow an additional `use_psa` parameter, calling `wrap_key_as_opaque()` if set, but otherwise keeping the code. All tests for `x509_csr_check()` can then be duplicated once with the new `use_psa` parameter set.

Reply:
On a general note, could we please at least briefly document such design considerations and discussions somewhere? It makes review so much easier if one sees not only what has been done but also why, and in particular what has been considered and deliberately refused as an alternative, especially if it's the canonical approach.

Or is it documented somewhere and I just missed it again?

Reply:
> And why don't we have these problems for the ordinary CSR tests, too?

For the ordinary CRS test, we pass a `f_rng` argument to `mbedtls_x509write_csr_pem()` that's fully deterministic, and consider it enough to make the result fully deterministic. IMO that's fragile (imagine we change how we use the RNG for blinding or whatever in a way that affects what part of the RNG output is used for making the signature, or that we switch for deterministic ECDSA by default, or...) and I would do it differently if I was writing this test today, regardless of the next consideration.

The reason why this won't work with PSA is that it disregards `f_rng` and always uses its own. Getting to the internal PSA RNG and replacing it with a fixed one is technically possible but IMO poor design as it violates an abstract, in addition to the considerations above.

> Sorry for my ignorance, but why is the TBS part of the CSR random? 

The CSR includes a signature and we clearly want the signature to be verified, and not just the part being signed, as the signature is what we're modifying here. With ECDSA, the signature may be random. We shouldn't be making assumptions about whether the particular PSA implementation used does deterministic ECDSA by default or not.

> On a general note, could we please at least briefly document such design considerations and discussions somewhere? It makes review so much easier if one sees not only what has been done but also why, and in particular what has been considered and deliberately refused as an alternative, especially if it's the canonical approach.

I agree that kind of thing should be mentioned at least in the PR description, sometimes even in the commit message.

Reply:
@mpg Thanks for your explanations. I initially thought you'd say that the to-be-signed part itself was random, but apparently that was a misunderstanding.

Reply:
I'm sorry for not describing it properly - probably because I wanted the PR to be set up as fast as possible. I'll add that to the description. Thanks @mpg for explaining!